staging/src/k8s.io/api/storage/v1/types.go

```diff
@@ -102,3 +102,110 @@ const (
 	// binding will occur during Pod scheduing.
 	VolumeBindingWaitForFirstConsumer VolumeBindingMode = "WaitForFirstConsumer"
 )
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttachment captures the intent to attach or detach the specified volume
+// to/from the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired attach/detach volume behavior.
+	// Populated by the Kubernetes system.
+	Spec VolumeAttachmentSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status of the VolumeAttachment request.
+	// Populated by the entity completing the attach or detach
+	// operation, i.e. the external-attacher.
+	// +optional
+	Status VolumeAttachmentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```

Comment:
it doesn't appear this object follows normal spec/status conventions:

* there is no status subresource
* the rest handler allows any part of the object to be modified via the primary resource (typically, there would be a primary resource strategy which would clear status in PrepareForCreate and preserve existing status in PrepareForUpdate, and a status strategy which would do the opposite, only allowing status to be updated on update
* generation is not incremented on update when the spec changes

I also don't see this PR registering this resource to be served under `storage.k8s.io/v1`:

https://github.com/kubernetes/kubernetes/blob/0aba5577412ea9b2fd89a0c0843e9b0c94e8848d/pkg/registry/storage/rest/storage_storage.go#L63-L83

Reply:
@liggitt iiuc, we need to do something like this https://github.com/kubernetes/kubernetes/blob/0aba5577412ea9b2fd89a0c0843e9b0c94e8848d/pkg/registry/core/rest/storage_core.go#L191 to add the subresource?

Reply:
@bertinatto to help give you a head start, I went ahead and added the status subresource here: https://github.com/msau42/kubernetes/commit/e5ccf30d61b079d4a3d5a54777442139ad611187.  You can go ahead and pull it into your branch.

Unit tests are passing, but it still needs to be tested e2e.  I am running e2es now, which will test that current v1beta API did not break.

Reply:
Sorry, didn't get around to running the e2es, and I g2g.  The hostpath tests are run as part of pull, but GCE PD tests which really test VolumeAttachment are serial.  If you need help running them, ask @davidz627

Reply:
I ran e2e btw and they passing. But - since we are still using beta version of `VolumeAttachment` objects, we aren't really testing new type via e2e yet.